File: src/bin/0053-maximum-subarray.py
# @leetup=info id=53 lang=python3 slug=maximum-subarray



# @leetup=code



class Solution:
    # def maxSubArray(self, nums: list[int]) -> int:

    #     @cache
    #     def doit(i: int, picked: bool) -> int:
    #         if i == len(nums):
    #             return 0 if picked else -inf
    #         if picked:
    #             # end or continue to pick
    #             return max(0, doit(i + 1, True) + nums[i])
    #         # skip or pick
    #         return max(doit(i + 1, False), doit(i + 1, True) + nums[i])

    #     return doit(0, False)

    def maxSubArray(self, nums: list[int]) -> int:
        N = len(nums)
        assert N
        dp = [0] * N
        dp[0] = nums[0]

        for i in range(1, N):
            dp[i] = max(nums[i], dp[i - 1] + nums[i])

        return max(dp)


# @leetup=code

File: src/bin/0053-maximum-subarray.cpp
// @leetup=info id=53 lang=cpp slug=maximum-subarray

#include <iostream>

// @leetup=code

#include <vector>
using namespace std;

class Solution {
 public:
  int maxSubArray(vector<int>& nums) {
    int best = nums[0];
    int pre = nums[0];
    for (int i = 1; i < nums.size(); ++i) {
      pre = max(nums[i], pre + nums[i]);
      best = max(best, pre);
    }
    return best;
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}
